### ISSUE TITLE
Stream-Stream join gives null results in addition to valid result

### ISSUE THREAD

**Issue by rmoff** [confluentinc/ksql#1558]
5.0.0-beta180702222458

Source data: 

```
ksql> SELECT TIMESTAMPTOSTRING(O.ROWTIME,'yyyy-MM-dd HH:mm:ss'), O.ORDER_ID FROM ORDERS O;
2018-07-07 23:50:17 | 1
2018-07-07 23:50:17 | 2
^CQuery terminated
ksql>
ksql> SELECT TIMESTAMPTOSTRING(OI.ROWTIME,'yyyy-MM-dd HH:mm:ss'), OI.ORDER_ID FROM ORDER_ITEMS OI;
2018-07-07 23:50:19 | 1
2018-07-07 23:50:19 | 1
2018-07-07 23:50:19 | 1
2018-07-07 23:50:19 | 2
```

Join gives results as expected:

```
ksql> SELECT O.ORDER_ID, O.TOTAL_AMOUNT, OI.PRODUCT_NAME FROM ORDERS O LEFT JOIN ORDER_ITEMS OI WITHIN 10 MINUTES ON O.ORDER_ID = OI.ORDER_ID;
1 | 10.5 | Foo
1 | 10.5 | Bar
1 | 10.5 | Baz
2 | 3.32 | Foo
```

No results, as expected: 

```
ksql> SELECT O.ORDER_ID, O.TOTAL_AMOUNT, OI.PRODUCT_NAME FROM ORDERS O LEFT JOIN ORDER_ITEMS OI WITHIN 2 SECOND ON O.ORDER_ID = OI.ORDER_ID;
1 | 10.5 | null
2 | 3.32 | null
^CQuery terminated
```

Running same query but set `WITHIN` to three seconds get _both_ null and the correct results: 

```
ksql> SELECT O.ORDER_ID, O.TOTAL_AMOUNT, OI.PRODUCT_NAME FROM ORDERS O LEFT JOIN ORDER_ITEMS OI WITHIN 3 SECOND ON O.ORDER_ID = OI.ORDER_ID;
1 | 10.5 | null
2 | 3.32 | null
1 | 10.5 | Foo
1 | 10.5 | Bar
1 | 10.5 | Baz
2 | 3.32 | Foo
^CQuery terminated
```

Re-running the exact query straight after doesn't repeat this: 

```
ksql> SELECT O.ORDER_ID, O.TOTAL_AMOUNT, OI.PRODUCT_NAME FROM ORDERS O LEFT JOIN ORDER_ITEMS OI WITHIN 3 SECOND ON O.ORDER_ID = OI.ORDER_ID;
1 | 10.5 | Foo
1 | 10.5 | Bar
1 | 10.5 | Baz
2 | 3.32 | Foo
```

**Comment by rmoff** [comment]
Seeing this again, with different datasets: 

```
ksql> SELECT ORDER_ID, TOTAL_AMOUNT, CUSTOMER_NAME FROM ORDERS;
1 | 10.5 | Bob Smith
2 | 3.32 | Sarah Black
3 | 21.0 | Emma Turner
^CQuery terminated
ksql>
ksql> SELECT ORDER_ID, SHIPMENT_ID, WAREHOUSE FROM SHIPMENTS;
1 | 42 | Nashville
3 | 43 | Palo Alto
^CQuery terminated
ksql> SELECT O.ORDER_ID, O.TOTAL_AMOUNT, O.CUSTOMER_NAME, S.SHIPMENT_ID, S.WAREHOUSE FROM ORDERS O LEFT JOIN SHIPMENTS S WITHIN 1 HOURS ON O.ORDER_ID = S.ORDER_ID;
1 | 10.5 | Bob Smith | null | null
2 | 3.32 | Sarah Black | null | null
3 | 21.0 | Emma Turner | null | null
1 | 10.5 | Bob Smith | 42 | Nashville
3 | 21.0 | Emma Turner | 43 | Palo Alto
^CQuery terminated
```

**Comment by rmoff** [comment]
the `WITHIN` seems to make a difference: 

```
ksql> SELECT O.ORDER_ID, O.TOTAL_AMOUNT, O.CUSTOMER_NAME, S.SHIPMENT_ID, S.WAREHOUSE FROM ORDERS O LEFT JOIN SHIPMENTS S WITHIN 8 HOURS ON O.ORDER_ID = S.ORDER_ID;
1 | 10.5 | Bob Smith | 42 | Nashville
2 | 3.32 | Sarah Black | null | null
3 | 21.0 | Emma Turner | 43 | Palo Alto
^CQuery terminated
ksql> SELECT O.ORDER_ID, O.TOTAL_AMOUNT, O.CUSTOMER_NAME, S.SHIPMENT_ID, S.WAREHOUSE FROM ORDERS O LEFT JOIN SHIPMENTS S WITHIN 1 HOURS ON O.ORDER_ID = S.ORDER_ID;
1 | 10.5 | Bob Smith | null | null
2 | 3.32 | Sarah Black | null | null
3 | 21.0 | Emma Turner | null | null
1 | 10.5 | Bob Smith | 42 | Nashville
3 | 21.0 | Emma Turner | 43 | Palo Alto
```

**Comment by rmoff** [comment]
Timestamps for the messages: 

```
ksql> SELECT ROWTIME, ROWKEY, ORDER_ID, TOTAL_AMOUNT, CUSTOMER_NAME FROM NEW_ORDERS;
1532000619440 | 1 | 1 | 10.5 | Bob Smith
1532000619448 | 2 | 2 | 3.32 | Sarah Black
1532000619449 | 3 | 3 | 21.0 | Emma Turner
^CQuery terminated
ksql> SELECT ROWTIME, ROWKEY, ORDER_ID, SHIPMENT_ID, WAREHOUSE FROM SHIPMENTS;
1532000801389 | 1 | 1 | 42 | Nashville
1532000801396 | 3 | 3 | 43 | Palo Alto
```

**Comment by apurvam** [comment]
Hi, @rmoff . This is actually the expected behavior. For the `LEFT JOIN`, only the left side stream triggers the join. The presence of `null` on the right side indicates that the left side event arrived first, and triggered the join with a `null` right. When the right hand side event arrives, the join result is updated and emits a non-null record. 

If you  want both sides to trigger the join (and hence never get null joins), you should use an `INNER` join.

**Comment by apurvam** [comment]
This is roughly similar to this: https://github.com/confluentinc/ksql/issues/1510

**Comment by rmoff** [comment]
But with the same set of data, I get different results running the same query - is that expected? 

And why would [varying `WITHIN`](https://github.com/confluentinc/ksql/issues/1558#issuecomment-404680543) make a difference?

**Comment by apurvam** [comment]
I think the `WITHIN` is a red herring. Running the same query multiple times would result in different results with a `LEFT JOIN` because it depends on the order in which the left and right streams are consumed. This is non-deterministic right now, and hence the results are non-deterministic.

**Comment by mjsax** [comment]
I don't think this can be fixed at KSQL level. This is how Kafka Streams works and KSQL inherits this behavior. It's not easy to change though.

**Comment by mjsax** [comment]
Fixed in Apache Kafka 3.0 (https://issues.apache.org/jira/browse/KAFKA-10847) and shipped in kslqDB 0.20 (https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-36-grace-stream-stream-joins.md)
